# HG changeset patch
# User Marco Caliari <marco.caliari@univr.it>
# Date 1525358850 -7200
#      Thu May 03 16:47:30 2018 +0200
# Branch stable
# Node ID a8a4b6e6e754479c838cb3c294f23d5b8fe96b59
# Parent  087030cf86a03eea5edddb5527a1ccdb3523dfbf
Fix lrwork dimension in svd for complex cases (bug #53823).

* svd.cc: Due to a bug in LAPACK <= 3.6, if jobz == 'N' then lrwork must be set
to 7*min_mn.

diff --git a/liboctave/numeric/svd.cc b/liboctave/numeric/svd.cc
--- a/liboctave/numeric/svd.cc
+++ b/liboctave/numeric/svd.cc
@@ -217,22 +217,23 @@ namespace octave
     svd<ComplexMatrix>::gesdd (char& jobz, F77_INT m, F77_INT n,
                                Complex *tmp_data, F77_INT m1, double *s_vec,
                                Complex *u, Complex *vt, F77_INT nrow_vt1,
                                std::vector<Complex>& work, F77_INT& lwork,
                                F77_INT *iwork, F77_INT& info)
     {
 
       F77_INT min_mn = std::min (m, n);
+      F77_INT max_mn = std::max (m, n);
 
       F77_INT lrwork;
       if (jobz == 'N')
         lrwork = 7*min_mn;
       else
-        lrwork = 5*min_mn*min_mn + 5*min_mn;
+        lrwork = min_mn * std::max (5*min_mn+5, 2*max_mn+2*min_mn+1);
 
       std::vector<double> rwork (lrwork);
 
       GESDD_COMPLEX_STEP (zgesdd, ZGESDD, F77_DBLE_CMPLX_ARG);
 
       lwork = static_cast<F77_INT> (work[0].real ());
       work.reserve (lwork);
 
@@ -250,19 +251,19 @@ namespace octave
                                     F77_INT& lwork, F77_INT *iwork,
                                     F77_INT& info)
     {
       F77_INT min_mn = std::min (m, n);
       F77_INT max_mn = std::max (m, n);
 
       F77_INT lrwork;
       if (jobz == 'N')
-        lrwork = 5*min_mn;
+        lrwork = 7*min_mn;
       else
-        lrwork = min_mn * std::max (5*min_mn+7, 2*max_mn+2*min_mn+1);
+        lrwork = min_mn * std::max (5*min_mn+5, 2*max_mn+2*min_mn+1);
       std::vector<float> rwork (lrwork);
 
       GESDD_COMPLEX_STEP (cgesdd, CGESDD, F77_CMPLX_ARG);
 
       lwork = static_cast<F77_INT> (work[0].real ());
       work.reserve (lwork);
 
       GESDD_COMPLEX_STEP (cgesdd, CGESDD, F77_CMPLX_ARG);
